Improvements related to layouts and some other things

sunsetsystems opened the pull request:
This is kind of a big PR as the things in it are inter-related and have evolved over time.

The checkout page for IPPF vs. normal use are separate files because the two are very different. It was done with two commits, one that renames the original pos_checkout.php and another that replaces it with a control script. This was the best way I could figure out to make sure git will sensibly track past and ongoing changes to both files.

#### Short description of what this resolves:

Added a version of the checkout page specific to IPPF.
Added ability to implement multiple layouts for patient history.
Added a layout condition to either hide a field or set it to a specified value.
Option "E" to allow multiple billing codes in a layout field.
Improvements for LBF static text fields.
Added LBF data type for a set of lab test results.
Added LBF data type to specify a patient.
Added NPI to the address book list.
Added "Save and Close Visit" button to LBFs when the visit has no charges.
Added option to display chosen layout groups as initially open.
Added option to show referrals in layout based forms.
Tracking and display of the date that a layout was last updated.
Added ability to deactivate and re-activate layouts.
Added ability to copy layout fields to another layout.
Other improvements to the layout editor.
Other assorted bug fixes.

bradymiller commented on interface/forms/LBF/new.php:
```diff
@@ -586,21 +608,53 @@ function capitalizeMe(elem) {
+                    var tmpId = w.name;
+                    if (tmpId.indexOf('enctabs-') == 0 && tmpId != window.name) {
+                        if (typeof w.somethingChanged !== 'undefined' && w.somethingChanged) {
+                            alert('<?php echo xla('Hold on! You have unsaved changes in another form. Please just Save this form and then complete the other one.'); ?>');
```
```
alert(<?php echo xlj('Hold on! You have unsaved changes in another form. Please just Save this form and then complete the other one.'); ?>);
```
(note removal of the outside quotes)

bradymiller commented on interface/forms/LBF/new.php:
```diff
@@ -586,21 +608,53 @@ function capitalizeMe(elem) {
+        function openLBFEncounterForm(formdir, formname, formid) {
+            top.restoreSession();
+            var url = '<?php echo "$rootdir/patient_file/encounter/view_form.php?formname=" ?>' +
+                formdir + '&id=' + formid;
```
```
encodeURIComponent(formdir) + '&id=' + encodeURIComponent(formid);
```

bradymiller commented on interface/forms/LBF/new.php:
```diff
@@ -586,21 +608,53 @@ function capitalizeMe(elem) {
+
+        function openLBFNewForm(formdir, formname) {
+            top.restoreSession();
+            var url = '<?php echo "$rootdir/patient_file/encounter/load_form.php?formname=" ?>' + formdir;
```
```
var url = '<?php echo "$rootdir/patient_file/encounter/load_form.php?formname=" ?>' + encodeURIComponent(formdir);
```

bradymiller commented on interface/forms/LBF/new.php:
```diff
@@ -1531,6 +1589,94 @@ function warehouse_changed(sel) {
+                    "WHERE " .
+                    "f.pid = ? AND f.encounter = ? AND f.formdir = 'LBFref' AND f.deleted = 0 " .
+                    "ORDER BY refer_date, f.form_id",
+                    array($pid, $encounter)
```
dang, now that is a sql query :)

![2DV (1)](https://user-images.githubusercontent.com/278968/112104332-4bd45b80-8b68-11eb-9021-eee0defa668d.gif)

bradymiller commented on interface/forms/LBF/new.php:
```diff
@@ -1598,6 +1756,9 @@ function warehouse_changed(sel) {
         </div>
     </div>
+
+    <p style='text-align:center' class='small'>Rev. <?php echo substr($grp_last_update, 0, 10); ?></p>
```
not sure where this variable comes from, but would be safe and surround it with a text()

sunsetsystems replied:
Yeah it comes from layout_group_properties.grp_last_update. Thanks.

bradymiller commented on interface/forms/LBF/printable.php:
```diff
@@ -616,6 +618,8 @@ function getContent()
 
 ?>
 
+<p style='text-align:center' class='small'>Rev. <?php echo substr($grp_last_update, 0, 10); ?></p>
```
not sure where this variable comes from, but would be safe and surround it with a text()

bradymiller commented on interface/super/edit_layout.php:
```diff
@@ -459,15 +447,51 @@ function encodeModifier($jsonArray)
+            array($layout_id, $onefield)
+        );
+        if (empty($srow)) {
+            die("Internal error: Field '$onefield' not found in layout '$layout_id'.");
```
escape the 2 variables above with text()

bradymiller commented on interface/super/edit_layout.php:
```diff
@@ -459,15 +447,51 @@ function encodeModifier($jsonArray)
+            array($tlayout, $onefield)
+        );
+        if (!empty($trow)) {
+            echo "<!-- Field '$onefield' already exists in layout '$tlayout'. -->\n";
```
escape the 2 variables above with text()

bradymiller commented on interface/super/edit_layout.php:
```diff
@@ -1454,30 +1490,29 @@ function myChangeCheck() {
             echo "&nbsp; &nbsp; ";
-            echo " <input type='button' class='deletegroup' id='" . attr($group_id) . "' value='" . xla('Delete Group') . "'/>";
-            ******************************************************************/
+            echo " <input type='button' class='deletegroup' id='$group_id' value='" . xl('Delete Group') . "'/>";
```
the id variables should be run through attr() on the  code block. also need xla() above line and below line value labels

bradymiller commented on interface/super/edit_layout.php:
```diff
@@ -1454,30 +1490,29 @@ function myChangeCheck() {
+            echo " <input type='button' class='movegroup' id='$group_id~down' value='" . xl('Move Down') . "'/>";
             echo "&nbsp; &nbsp; ";
-            echo "<input type='button' class='btn btn-secondary btn-sm' value='" . xla('Group Properties') . "' onclick='edit_layout_props(" . attr_js($group_id) . ")' />";
+            echo "<input type='button' value='" . xla('Group Properties') . "' onclick='edit_layout_props(\"$group_id\")' />";
```
the onlclick portion can be escaped by:
```
onclick='edit_layout_props(" . attr_js($group_id) . ")'
```

bradymiller commented on interface/super/edit_layout.php:
```diff
@@ -1969,28 +1925,22 @@ function validateNewField(idpfx) {
-        var groupname = parts.replace(/^\d+/, "");
-        if (confirm(<?php echo xlj('WARNING') ?> + " - " + <?php echo xlj('This action cannot be undone.') ?> + "\n" + <?php echo xlj('Are you sure you wish to delete the entire group named'); ?> + " '" + groupname + "'?")) {
+        if (confirm("<?php echo xls('WARNING') . ' - ' . xls('This action cannot be undone.') . '\n' .
+            xls('Are you sure you wish to delete this entire group?'); ?>")
```
should be using xlj instead of xls

sunsetsystems replied:
Hmm. xls() is part of a javascript string literal, so json_encode() output there will not be evaluated as an expression. Do you mean like this:

        if (confirm(<?php echo js_escape(xl('WARNING') . ' - ' . xl('This action cannot be undone.') . '\n' .
            xl('Are you sure you wish to delete this entire group?')); ?>)

... but I wonder what js_escape() will do to my '\n' which is intended to force a line break.

bradymiller replied:
Guessing that could cause issues. Aren't able to separate out the \n character like is done in above line 1986 that was removed?

sunsetsystems replied:
Yep I'll try that.

bradymiller commented on interface/super/edit_layout.php:
```diff
@@ -2187,6 +2128,25 @@ function layoutLook(){
+function typeUsesList(datatype) {
+<?php
+foreach ($typesUsingList as $mytype) {
+    echo "  if (datatype == '$mytype') return true;\n";
```
would escape $mytype with attr()

bradymiller commented on interface/super/edit_layout_props.php:
```diff
@@ -232,14 +244,27 @@ function get_related() {
+<?php if (empty($group_id) && !empty($row['grp_last_update'])) { ?>
+ <tr>
+  <td valign='top' width='1%' nowrap>
+    <?php echo xls('Last Update'); ?>
```
xlt

bradymiller commented on interface/super/edit_layout_props.php:
```diff
@@ -407,6 +432,17 @@ function get_related() {
+    <?php /* ?>
+ <tr>
+  <td valign='top' width='1%' nowrap>
+   <?php echo xls('Enable Save and Close'); ?>
```
xlt

bradymiller commented on interface/super/edit_layout_props.php:
```diff
@@ -437,6 +473,36 @@ function get_related() {
+<?php } else { // else this is a group ?>
+ <tr>
+  <td valign='top' width='1%' nowrap>
+    <?php echo xls('Initially Open Group'); ?>
```
xlt

bradymiller commented on library/options.inc.php:
```diff
@@ -172,55 +181,66 @@ function generate_select_list(
-            $got_selected = true;
+            // Already has been translated above (if applicable), so do not need to use
+            // the xl_list_label() function here
+            $optionLabel = text(xl_list_label($lrow ['title']));
```
this change looks like it is ignoring the prior comment

bradymiller commented on library/options.inc.php:
```diff
@@ -294,28 +314,179 @@ function generate_select_list(
+    } else {
+        $s .= "<input type='text'";
+        if ($outtype == 0) {
+            $s .= " name='$name' id='$name'";
```
attr() above 2 variables

bradymiller commented on library/options.inc.php:
```diff
@@ -294,28 +314,179 @@ function generate_select_list(
+        if ($outtype == 0) {
+            $s .= " name='$name' id='$name'";
+        }
+        $s .= " size='$size' $string_maxlength" .
```
attr above $size variable

bradymiller commented on library/options.inc.php:
```diff
@@ -294,28 +314,179 @@ function generate_select_list(
+                $s .= xl_list_label($lrow['title']);
+                $s .= ':' . $avalue[$option_id][0];
+                $s .= ':' . $avalue[$option_id][2];
+                $s .= ':' . $avalue[$option_id][3];
```
looks like all 4 variables above need to be escaped by text()

bradymiller commented on library/options.inc.php:
```diff
@@ -716,25 +895,20 @@ function generate_form_field($frow, $currvalue)
-                    } else {
-                        $currdescstring .= $codestring;
-                    }
+                    $currdescstring .= getCodeDescription($codestring, $codetype);
```
i don't think this function is looking up codes correctly (ie. just looking at codes table instead of what lookup_code_descriptions() does which is supports both internal and external code types

sunsetsystems replied:
OK I'm changing getCodeDescription() to call lookup_code_descriptions(). Thanks.

bradymiller commented on library/options.inc.php:
```diff
@@ -716,25 +895,20 @@ function generate_form_field($frow, $currvalue)
             echo "<input type='text'" .
             " name='form_$field_id_esc'" .
             " id='form_related_code'" .
+            " class='$className'" .
```
attr()
(yep, i am html escaping crazy :) )

bradymiller commented on library/options.inc.php:
```diff
@@ -751,10 +925,12 @@ function generate_form_field($frow, $currvalue)
             echo "<input type='text'" .
             " name='form_$field_id_esc'" .
             " id='form_related_code'" .
+            " class='$className'" .
```
attr()

bradymiller commented on library/options.inc.php:
```diff
@@ -1454,7 +1664,10 @@ function generate_print_field($frow, $currvalue)
     // generic single-selection list
     //  Supports backup lists.
-    if (false && ($data_type == 1 || $data_type == 26 || $data_type == 43 || $data_type == 46)) {
+    // if (false && ($data_type == 1 || $data_type == 26 || $data_type == 33 || $data_type == 43)) {
```
note this comment is missing 46

bradymiller commented on library/options.inc.php:
```diff
@@ -1893,7 +2070,7 @@ function generate_print_field($frow, $currvalue)
 
         echo "</table>";
-    } elseif ($data_type == 27 || $data_type == 1 || $data_type == 26 || $data_type == 46) {
+    } elseif ($data_type == 27 || $data_type == 46) { // Removed: || $data_type == 1 || $data_type == 26 || $data_type == 33
```
i think 46 should also go since now captured above (i always wondered why that `false` thing was there)

sunsetsystems replied:
false was there because I anticipated re-using the code. :)

bradymiller commented on library/options.inc.php:
```diff
@@ -3032,13 +3236,42 @@ function accumActionConditions($field_id, &$condition_str, &$condarr)
+            // For billing codes handle requirement to display its description.
+            $tmp = explode('=', $action, 2);
+            if (!empty($tmp[1])) {
+                $condition_str .= "valdesc:'" . addslashes(getCodeDescription($tmp[1])) . "', ";
```
does the  following work (much more robust js escaping):
```
$condition_str .= "valdesc:" . js_escape(getCodeDescription($tmp[1])) . ", ";
```
(note dropped the outer quotes)

sunsetsystems replied:
I think so, thanks.

bradymiller commented on library/options.inc.php:
```diff
@@ -3032,13 +3236,42 @@ function accumActionConditions($field_id, &$condition_str, &$condarr)
+                $condition_str .= "valdesc:'" . addslashes(getCodeDescription($tmp[1])) . "', ";
+            }
+        }
+        $condition_str .= "andor:'" . js_escape($andor) . "'}";
```
this is likely broken since there are outer quotes. rec removing the outer quotes:
```
$condition_str .= "andor:" . js_escape($andor) . "}";
```

bradymiller commented on library/options.inc.php:
```diff
@@ -3032,13 +3236,42 @@ function accumActionConditions($field_id, &$condition_str, &$condarr)
+    }
+}
+
+function getCodeDescription($codestring, $defaulttype = 'ICD10')
```
should instead base this function on lookup_code_descriptions()

bradymiller commented on library/options.inc.php:
```diff
@@ -3461,56 +3780,167 @@ function display_layout_tabs_data($formtype, $result1, $result2 = '')
+                        echo "<br /><span class='bold'><input type='checkbox' name='form_cb_" .
+                            attr($gs_group_seq) . "' value='1' " .
+                            "onclick='return divclick(this,\"div_" .
+                            attr(addslashes($gs_group_seq)) . "\");'";
```
try:
```
"onclick='return divclick(this," . js_escape("div_" .  attr($gs_group_seq)) . ");'"
```

bradymiller commented on sql/6_0_0-to-6_1_0_upgrade.sql:
```diff
@@ -453,4 +453,25 @@ SET @codetypeid = (SELECT `ct_id` FROM `code_types` WHERE `ct_key` = 'CVX');
+#EndIf
+
+#IfMissingColumn layout_group_properties grp_last_update
+ALTER TABLE `layout_group_properties` ADD COLUMN `grp_last_update` timestamp;
```
timestamp NULL
(something we are doing now to avoid blank/bad date/timestamps)

bradymiller commented on sql/database.sql:
```diff
@@ -3212,9 +3212,13 @@ CREATE TABLE `layout_group_properties` (
   grp_services    varchar(4095)  not null default '',
   grp_products    varchar(4095)  not null default '',
   grp_diags       varchar(4095)  not null default '',
+  grp_last_update timestamp,
```
timestamp NULL

bradymiller commented:
review almost complete. just gotta look through the interface/patient_file/pos_checkout_ippf.php file (will do tomorrow)

bradymiller commented:
oops, re-opened

bradymiller commented on interface/patient_file/pos_checkout_normal.php:
```diff
@@ -0,0 +1,1158 @@
+        <title><?php echo xlt('Receipt for Payment'); ?></title>
+        <script>
+
+        <?php require($GLOBALS['srcdir'] . "/restoreSession.php"); ?>
```
@sunsetsystems , This type of thing has always confused me. Shouldn't this script be only in `top` since even popups look for it in `top`?

sunsetsystems replied:
Yes. Long ago I put this in dialogs but later realized that was unnecessary.

bradymiller commented on interface/patient_file/pos_checkout_normal.php:
```diff
@@ -0,0 +1,1158 @@
+        <script>
+            var mypcc = <?php echo js_escape($GLOBALS['phone_country_code']); ?>;
+
+            <?php require($GLOBALS['srcdir'] . "/restoreSession.php"); ?>
```
same question as above (ie. why are popups including this)?
(note this is not really part of this code review since existing code; just trying to figure out why some of the popups are doing this)

sjpadgett replied:
Sometimes top is not available or in scope. I'm thinking we could add a dependency check in utility then include if needed.

sunsetsystems replied:
Yeah. Guess I'll leave this alone for now, but at some point it should be changed and tested.

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+         text(oeFormatShortDate($postdate)) . "</td>\n";
+    echo "  <td>$code</td>\n";
+    echo "  <td>$description</td>\n";
+    echo "  <td class='text-center'>" . ($isadjust ? '' : $quantity) . "</td>\n";
```
text() on $code, $description, $quantity and around all the oeFormatMoney() calls below

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        echo " title='" . xla('Entered') . ' ' .
+            text(oeFormatShortDate($billtime)) . attr(substr($billtime, 10)) . "'";
+    }
+    echo ">" . oeFormatShortDate($paydate) . "</td>\n";
```
text()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+    echo "  <td colspan='2'>" . text($refno) . "</td>\n";
+    echo "  <td colspan='$rcpt_num_method_columns' class='text-left'>" . text($method) . "</td>\n";
+    echo "  <td colspan='$rcpt_num_ref_columns' class='text-left'>" . text($description) . "</td>\n";
+    echo "  <td colspan='$rcpt_num_amount_columns' class='text-right'>" . oeFormatMoney($amount) . "</td>\n";
```
text()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        );
+    }
+    if (empty($ferow)) {
+        die(xl("This patient has no activity."));
```
xlt()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+    function voidwrap(form_reason, form_notes) {
+        top.restoreSession();
+        document.location.href = 'pos_checkout.php?ptid=<?php echo $patient_id; ?>&' + voidaction +
+          '=<?php echo $encounter; ?>' + '&form_checksum=<?php echo $current_checksum; ?>' +
```
document.location.href = 'pos_checkout.php?ptid=' + <?php echo js_url($patient_id); ?> + '&' + encodeURIComponent(voidaction) +
          '=' + <?php echo js_url($encounter); ?> + '&form_checksum=' + <?php echo js_url($current_checksum); ?> +

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+    function voidme(action) {
+        voidaction = action;
+        if (action == 'void' || action == 'voidall') {
+           if (!confirm('<?php echo xl('This will advance the receipt number. Please print the receipt if you have not already done so.'); ?>')) {
```
xlj()
(ensure drop outer quotes)

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+    <?php
+    echo xlt("Client Receipt");
+    if ($invoice_refno) {
+        echo " " . xlt("for Invoice") . " $invoice_refno";
```
text() on $invoice_refno

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+                    <tr>
+                        <td width='50%' class='text-left' valign='top'>
+                            <?php echo $patdata['fname'] . ' ' . $patdata['mname'] . ' ' . $patdata['lname']; ?>
+                            <br /><?php echo $patdata['street']; ?>
```
text() above 2 lines

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+                            <br /><?php
+                            echo generate_layout_display_field('DEM', 'city', $patdata['city']) . ", ";
+                            echo generate_layout_display_field('DEM', 'state', $patdata['state']) . " ";
+                            echo $patdata['postal_code']; ?>
```
text()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+                                </tr>
+                                <tr>
+                                    <td><?php echo xlt('Ending Account Balance'); ?>&nbsp;&nbsp;&nbsp;&nbsp;</td>
+                                    <td class='text-right'><?php echo oeFormatMoney($head_endbal); ?></td>
```
text() around all above oeFormatMoney() calls

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+                    <tr>
+                        <td colspan='<?php echo 6 + $num_optional_columns; ?>'
+                        style='padding-top:5pt;' class='font-weight-bold'>
+                                <?php echo xl('Charges for') . ' ' . oeFormatShortDate($svcdate); ?>
```
xlt() and text()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        }
+        echo "  <td align='right'>" . oeFormatMoney($aTotals[4]) . "</td>\n";
+        echo " </tr>\n";
+    }
```
above code block add xlt and/or text where needed

bradymiller replied:
For rest of this script, not gonna mention xlt() or text() stuff

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+                        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
+
+    <?php if ($details) { ?>
+                        <a href='pos_checkout.php?details=0&ptid=<?php echo $patient_id; ?>&enc=<?php echo attr($encounter); ?>' onclick='top.restoreSession()'><?php echo xlt('Hide Details'); ?></a>
```
```
<a href='pos_checkout.php?details=0&ptid=<?php echo attr_url($patient_id); ?>&enc=<?php echo attr_url($encounter); ?>' onclick='top.restoreSession()'><?php echo xlt('Hide Details'); ?></a>
```

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+    <?php if ($details) { ?>
+                        <a href='pos_checkout.php?details=0&ptid=<?php echo $patient_id; ?>&enc=<?php echo attr($encounter); ?>' onclick='top.restoreSession()'><?php echo xlt('Hide Details'); ?></a>
+    <?php } else { ?>
+                        <a href='pos_checkout.php?details=1&ptid=<?php echo $patient_id; ?>&enc=<?php echo attr($encounter); ?>' onclick='top.restoreSession()'><?php echo xlt('Show Details'); ?></a>
```
```
<a href='pos_checkout.php?details=1&ptid=<?php echo attr_url($patient_id); ?>&enc=<?php echo attr_url($encounter); ?>' onclick='top.restoreSession()'><?php echo xlt('Show Details'); ?></a>
```

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+    <script>
+    <?php
+    if ($alertmsg) {
+        echo " alert('" . addslashes($alertmsg) . "');\n";
```
js_escape()
(and drop outer quotes)

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+    echo "<input type='hidden' name='line[$lino][billtime]' value='$billtime'>";
+    echo "</td>\n";
+    echo "  <td class='text'>$description</td>";
+    echo "  <td class='text' align='right'>$units</td>";
```
ensure attr() and text()
(gonna stop mentioning attr() on this script review)

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        "list_id = 'adjreason' AND option_value = 1 AND activity = 1"
+    );
+    while ($tmprow = sqlFetchArray($tmpres)) {
+        echo "            && adjreason != '" . addslashes($tmprow['option_id']) . "'\n";
```
use js_escape() instead
(ensure remove outer quotes)

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+            var totaltax = parseFloat(f['totaltax[' + i + ']'].value) + tax;
+            f['totaltax[' + i + ']'].value = totaltax.toFixed(<?php echo $currdecimals ?>);
+        }
+        f[pfx + '[amount]'].value = extended.toFixed(<?php echo $currdecimals ?>);
```
escape above $currdecimals with js_escape()

sunsetsystems replied:
I added code to force it to be an integer.

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+            amount = parseFloat(amount.toFixed(<?php echo $currdecimals ?>));
+            total += amount;
+            // Set payment row's description to Refund if the amount is negative.
+            var title = amount < 0 ? '<?php echo xl('Refund'); ?>' : '<?php echo xl('New payment'); ?>';
```
xlj() and drop outer quotes

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+            preg_match('/ADJ=(\w+)/', $tmprow['notes'], $matches) ||
+            preg_match('/ADJ="(.*?)"/', $tmprow['notes'], $matches)
+        ) {
+            echo "  if (customer == '" . addslashes($tmprow['option_id']) . "') ret = '" . addslashes($matches[1]) . "';\n";
```
use js_escape() and drop outer quotes

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+                url: "<?php echo $GLOBALS['webroot']; ?>/library/ajax/check_szf_referrals_ajax.php",
+                data: {
+                    "pid": <?php echo $patient_id; ?>,
+                    "encounter": <?php echo $encounter_id; ?>,
```
js_escape() or both lines above

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        "list_id = 'paymethod' AND activity = 1 AND notes LIKE '%GIFTCARD=Y%' ORDER BY seq, title"
+    );
+    while ($lrow = sqlFetchArray($lres)) {
+        echo "                || method == '" . addslashes($lrow['option_id']) . "'\n";
```
use js_escape() instead and drop outer quotes

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        }
+<?php } else { ?>
+        if (missingtypeamt) {
+            alert('<?php echo xls('Adjustment type is required for each line with an adjustment.') ?>');
```
xlj in 4 lines in above code block

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+<body>
+
+<?php
+echo "<form method='post' action='pos_checkout.php?rde=$rapid_data_entry";
```
```
echo "<form method='post' action='pos_checkout.php?rde=<?php echo attr_url($rapid_data_entry); ?>";
```

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+<?php
+echo "<form method='post' action='pos_checkout.php?rde=$rapid_data_entry";
+if ($encounter_id) {
+    echo "&enid=$encounter_id";
```
attr_url()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+            $query .= "(modifier IS NULL OR modifier = '')";
+        }
+        $query .= " LIMIT 1";
+        $tmp = sqlQuery($query);
```
need to use sql binding in this query

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+<?php
+// A current invoice reference number may be present if there was a previous checkout.
+$tmprow = sqlQuery("SELECT invoice_refno FROM form_encounter WHERE " .
+  "pid = '$patient_id' AND encounter = '$encounter_id'");
```
need to use sql binding in this query

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        echo "    cell.colSpan = $form_num_ref_columns;\n";
+    }
+    if ($ix == 3) {
+        echo "    cell.colSpan = $form_num_amount_columns;\n";
```
is these are going out to javascript context, then would wrap the variables with js_escape()

sunsetsystems replied:
These are hard coded to integer values.

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+ function payprevious() {
+  var width  = 750;
+  var height = 550;
+  var loc = '../patient_file/front_payment.php?omitenc=<?php echo $encounter_id; ?>';
```
```
var loc = '../patient_file/front_payment.php?omitenc=' + <?php echo js_url($encounter_id); ?>;
```

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+
+<?php
+if ($alertmsg) {
+    echo "alert('" . addslashes($alertmsg) . "');\n";
```
js_escape and drop outer quotes

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+                array($patient_id, $encounter_id)
+            );
+            if (empty($grow['count'])) { // if there is no gcac form
+                echo " alert('" . xl('This visit will need a GCAC form, referral or procedure service.') . "');\n";
```
xlj and drop outer quotes

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2798 @@
+        } else {
+            echo xls('new method in contraception form does not match the contraception service');
+        }
+        echo "');\n";
```
another challenging one :) lets try:
```
        $warningMessage = xl('Warning') . ': ';
        if (!$csmethod) {
            $warningMessage .= xl('there is a contraception service but no contraception form new method');
        } else if (!$contraception_billing_code) {
            $warningMessage .= xl('there is a contraception form new method but no contraception service');
        } else {
            $warningMessage .= xl('new method in contraception form does not match the contraception service');
        }
         echo " alert(" . js_escape($warningMessage) . ");\n";
```

sunsetsystems replied:
Perfect, thanks!

bradymiller commented:
review complete :)

sunsetsystems commented:
Awesome review! I'll have an update sometime tomorrow your time.

sunsetsystems commented:
Fixes are done, I hope!

sunsetsystems commented:
I found some more things to fix. New commit coming shortly.

sjpadgett commented:
@sunsetsystems 
I have two small fixes to new.php and option.inc.php where i'm getting ready to put up a PR for portal and will result in a conflict.
They'd be minor, two or three lines that fix the group title and checkbox to be cleaner and adds replace container with container-fluid in new.php for portal use. The option.inc.php change is for the Exclude in Portal option and is 4 lines.

sunsetsystems commented:
@sjpadgett Sounds good, do you need me to do anything?

sjpadgett commented:
Nope, just letting ya know you'll probably get a conflict on those two files depending on who gets merged first. wanta race.:)

sunsetsystems commented:
Testing IPPF version of the checkout uncovered some bugs. I think they're all fixed now.

sunsetsystems commented:
I think what I want to do is squash all the commits together except for the first one, ending up with 2 commits for this PR. Notice if you look at just the first commit it shows pos_checkout.php renamed to pos_checkout_normal.php (which it was), whereas if you look at the set of files changed for all commits it looks like pos_checkout.php was replaced and not renamed, and like pos_checkout_normal.php is a brand new file.

This is to retain the history of what actually happened, but I'm not a git expert. @bradymiller do you agree that this is the right thing to do? I guess it means I need to push from the command line.

bradymiller commented:
@sunsetsystems , note you can do a forced push to your PR branch (ie. can rebase your commits (squash and deal with merge conflicts) and then something like `git push --force origin whrwork3` to push that to your PR).

sunsetsystems commented:
@bradymiller OK I've rebased, squashed and force-pushed to the original branch. Decided to go with 3 commits instead of 2. Let me know when you're OK with a merge.

bradymiller commented on interface/forms/LBF/new.php:
```diff
@@ -1534,6 +1593,94 @@ function warehouse_changed(sel) {
+                    }
+                    echo " <tr style='cursor:pointer;cursor:hand' " .
+                        "onclick=\"openLBFEncounterForm('LBFref', 'Referral', '" .
+                        $refrow['form_id'] . "')\">\n";
```
I missed one on last review :scream: 
```
"onclick=\"openLBFEncounterForm('LBFref', 'Referral', " . attr_js($refrow['form_id']) . ")\">\n";
```
(note outer quote removal)

bradymiller commented on interface/patient_file/void_dialog.php:
```diff
@@ -0,0 +1,102 @@
+ */
+
+$fake_register_globals = false;
+$sanitize_all_escapes  = true;
```
The above 2 horrific lines have been exiled from OpenEMR. :bomb: 
:)

bradymiller replied:
Come to think of it, they do have historical values. We could turn them into a OpenEMR legacy NFT and sell them to the highest bidder :)
https://opensea.io/accounts/OpenEMR

bradymiller commented on interface/patient_file/void_dialog.php:
```diff
@@ -0,0 +1,102 @@
+
+require_once('../globals.php');
+require_once("$srcdir/options.inc.php");
+require_once("$srcdir/htmlspecialchars.inc.php");
```
the above line is always included, so can remove

bradymiller commented on interface/patient_file/void_dialog.php:
```diff
@@ -0,0 +1,102 @@
+function DoSubmit() {
+  var f = document.forms[0];
+  if (opener.closed || !opener.voidwrap) {
+    alert('<?php echo xls('The destination form was closed; I cannot act on your selection.'); ?>');
```
xlj

bradymiller commented on interface/patient_file/void_dialog.php:
```diff
@@ -0,0 +1,102 @@
+    alert('<?php echo xls('The destination form was closed; I cannot act on your selection.'); ?>');
+  }
+  else if (f.form_reason.selectedIndex == 0) {
+    alert('<?php echo xls('Please specify Void Reason.'); ?>');
```
xlj

bradymiller commented on interface/patient_file/void_dialog.php:
```diff
@@ -0,0 +1,102 @@
+    return false;
+  }
+  else if (f.form_notes.value.trim() == '') {
+    alert('<?php echo xls('Please specify Void Notes.'); ?>');
```
xlj

bradymiller commented on interface/super/edit_layout.php:
```diff
@@ -1969,28 +1925,23 @@ function validateNewField(idpfx) {
-        if (confirm(<?php echo xlj('WARNING') ?> + " - " + <?php echo xlj('This action cannot be undone.') ?> + "\n" + <?php echo xlj('Are you sure you wish to delete the entire group named'); ?> + " '" + groupname + "'?")) {
+        if (confirm(<?php echo xlj('WARNING') ?> + " - " +
+            <?php echo xlj('This action cannot be undone.') ?> + "\n" +
+            <?php echo xlj('Are you sure you wish to delete this entire group?'); ?>)
```
![tenor99](https://user-images.githubusercontent.com/278968/112711794-4b99d000-8e88-11eb-9fe8-498f1491dcbd.gif)

bradymiller commented on library/checkout_receipt_array.inc.php:
```diff
@@ -0,0 +1,543 @@
+    } else {
+        $query .= "ORDER BY id DESC LIMIT 1";
+    }
+    $ferow = sqlQuery($query);
```
need to use sql binding in this query

bradymiller commented on library/checkout_receipt_array.inc.php:
```diff
@@ -0,0 +1,543 @@
+    }
+    $ferow = sqlQuery($query);
+    if (empty($ferow)) {
+        die(xl("This patient has no activity."));
```
xlt

bradymiller commented on library/checkout_receipt_array.inc.php:
```diff
@@ -0,0 +1,543 @@
+        "SELECT SUM(fee) AS amount FROM billing WHERE " .
+        "pid = ? AND encounter = ? AND activity = 1 AND " .
+        "code_type != 'COPAY'" .
+        ($billtime ? " AND billed = 1 AND bill_date <= '$billtime'" : ""),
```
sql binding needed here

bradymiller commented on library/checkout_receipt_array.inc.php:
```diff
@@ -0,0 +1,543 @@
+    $row = sqlQuery(
+        "SELECT SUM(fee) AS amount FROM drug_sales WHERE " .
+        "pid = ? AND encounter = ?" .
+        ($billtime ? " AND billed = 1 AND bill_date <= '$billtime'" : ""),
```
sql binding

bradymiller commented on library/checkout_receipt_array.inc.php:
```diff
@@ -0,0 +1,543 @@
+    $row = sqlQuery(
+        "SELECT SUM(pay_amount) AS payments, " .
+        "SUM(adj_amount) AS adjustments FROM ar_activity WHERE " .
+        "deleted IS NULL AND pid = '$patient_id' AND encounter = '$encounter'" .
```
sql binding

bradymiller commented on library/checkout_receipt_array.inc.php:
```diff
@@ -0,0 +1,543 @@
+        "SELECT SUM(fee) AS amount FROM billing WHERE " .
+        "pid = ? AND encounter = ? AND activity = 1 AND " .
+        "code_type = 'COPAY'" .
+        ($billtime ? " AND billed = 1 AND bill_date <= '$billtime'" : ""),
```
sql binding

bradymiller commented on library/checkout_receipt_array.inc.php:
```diff
@@ -0,0 +1,543 @@
+        " JOIN users AS u ON u.id = a.post_user" .
+        " WHERE" .
+        " a.deleted IS NULL AND a.pid = ? AND a.encounter = ?" .
+        ($billtime ? " AND a.post_time = '$billtime'" : "") .
```
sql binding

bradymiller commented on sql/database.sql:
```diff
@@ -10007,6 +10020,23 @@ INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `ac
 INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `activity`) VALUES ('page_validation', 'add_edit_event#theform_prov', '/interface/main/calendar/add_edit_event.php?prov=true', 170, '{}',  1);
 INSERT INTO `list_options` (`list_id`, `option_id`, `title`, `seq`, `notes`, `activity`) VALUES ('page_validation', 'messages#new_note','/interface/main/messages/messages.php',150, '{"form_datetime":{"futureDate":{"message": "Must be future date"}}, "reply_to":{"presence": {"message": "Please choose a patient"}}}', 1);
```
Does the shift list_options list entry need to be added? (it is in the ugprade script)

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2824 @@
+    function printme(checkout_id) {
+    <?php if (!empty($GLOBALS['gbl_custom_receipt'])) { ?>
+        // Custom checkout receipt needs to be sent as a PDF in a new window or tab.
+        window.open('pos_checkout.php?<?php echo "ptid=$patient_id&enc=$encounter&pdf=1&coid="; ?>' + encodeURIComponent(checkout_id),
```
```
window.open('pos_checkout.php?<?php echo "ptid=' + " . js_url($patient_id) . " + '&enc=' + " . js_url($encounter) . " + '&pdf=1&coid="; ?>' + encodeURIComponent(checkout_id),
```

bradymiller replied:
actually this is probably prettier:
```
window.open('pos_checkout.php?ptid=' + <?php echo js_url($patient_id); ?> + '&enc=' + <?php echo js_url($encounter); ?> + '&pdf=1&coid=' + encodeURIComponent(checkout_id),
```

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2824 @@
+    echo '&framed=1';
+}
+echo "' onsubmit='return validate()'>\n";
+echo "<input type='hidden' name='form_pid' value='$patient_id' />\n";
```
attr()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2824 @@
+echo "  <td class='bold' colspan='" . (!empty($GLOBALS['gbl_checkout_charges']) ? 4 : 3) .
+     "' align='right'>" . xlt('Total Charges This Visit') . "</td>\n";
+echo "  <td class='text' align='right'><input type='text' name='totalcba' " .
+     "value='$totalchg' size='6' maxlength='8' " .
```
attr()

bradymiller commented on interface/patient_file/pos_checkout_ippf.php:
```diff
@@ -0,0 +1,2824 @@
+            <?php echo xlt('Tentative Invoice Ref No'); ?>
+        </td>
+        <td class='text' align='right' colspan='<?php echo $form_num_amount_columns; ?>'>
+            <?php echo $irnumber; ?>
```
text()

bradymiller commented:
@sunsetsystems , Noted just a couple more issues above. After those are addressed, then feel free to merge it in :)

sunsetsystems commented:
OK gonna do another round of smoke-testing and then merge tomorrow. Thanks for all the help!
